gCNV restarts inference with a new random seed in case ELBO converges to a NaN

asmirnov239 opened the pull request:
This is intended to alleviate transient issues with GermlineCNVCaller inference in which the ELBO converges to a NaN value, by calling the python gCNV code with an updated random seed input.

asmirnov239 commented:
@cmnbroad I added the changes you suggested, could you take a look? 

I added a set of methods to `PythonScriptExecutor` mirroring existing ones that return booleans.  I tried to extract as much common code as possible. Also, because now the entry point in  `ScriptExecutor` that return `ProcessOutput` does not process and throw any exceptions, I extracted that processing code and manually call it from `GermlineCNVCaller`. Let me know if you have a better idea on a better way to handle it.

If that looks good to you, could you also suggest what unit tests you would like me to add?

cmnbroad reviewed (changes requested):
@asmirnov239 Thanks for the update. There are a bunch of minor/cosmetic comments, and one request for a minor refactoring change, but overall this looks pretty good.

As for tests, I think there are 4 `PythonScriptExecutorUnit` test cases:

- Two cases for `executeScriptAndGetOutput(final Resource scriptResource,..), one that executes a script that just does `exit 239`, and one that is some simple code that just succeeds without setting an explicit exit code, with both calling and validating the `ProcessOutput` exitcode, and the return value from getExceptionMessageFromScriptError`.
- The same two cases, only for `executeArgsAndGetOutput`.

Ideally there would also be a negative case for GermlineCNVCaller if you can force it to fail to converge, but maybe thats harder.

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -305,10 +310,30 @@ protected Object doWork() {
+            if (pythonProcessOutput.getExitValue() == gcnvRestartExitCode) {
+                final Random generator = new Random(randomGCNVSeed);
+                final int nextGCNVSeed = generator.nextInt();
+                logger.info("The inference ran into a NaN error and will be restarted one more time.");
```
Yes that's better I agree. Done

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/utils/python/PythonScriptExecutor.java:
```diff
@@ -105,12 +104,31 @@ public boolean executeModule(final String moduleName, final List<String> pythonP
      * @param scriptResource {@link Resource} for the script to execute
      * @param pythonProcessArgs args to be passed to the python process
-     * @param scriptArgs args to be passed to the python code
+     * @param scriptArgs args to be passed to the python codescriptArgs
```
Oops, this was an accident

asmirnov239 commented on src/main/resources/org/broadinstitute/hellbender/tools/copynumber/case_denoising_calling.py:
```diff
@@ -213,9 +214,14 @@ def update_args_dict_from_saved_model(input_model_path: str,
+        task.disengage()
+    except gcnvkernel.ConvergenceError as err:
+        logger.info(err.message)
+        sys.exit(239) # Pass an exit code to Java side indicating restart is needed
```
Definitely. I added it to gcnvkernel.io_const and kept it lower case to stick with the convention there

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/GermlineCNVCaller.java:
```diff
@@ -289,6 +291,9 @@
     private SimpleIntervalCollection specifiedIntervals;
     private File specifiedIntervalsFile;
 
```
Done!

asmirnov239 commented on src/main/java/org/broadinstitute/hellbender/utils/runtime/ScriptExecutor.java:
```diff
@@ -72,62 +74,52 @@ protected void executableMissing() {
-                logger.warn("Skipping: " + getApproximateCommandLine());
-                return false;
-            }
+    public String getExceptionMessageFromScriptError(final ProcessOutput po) {
```
I like that. Done!

asmirnov239 commented:
@cmnbroad I addressed your comments, except for making a negative test case for GermlineCNVCaller - I added that to  #6893 issue with an ongoing list of needed tests. Let me know if the changes look good to you.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/utils/python/PythonScriptExecutor.java:
```diff
@@ -205,4 +228,38 @@ public static void checkPythonEnvironmentForPackage(final String packageName) {
+    /**
+     * Auxiliary method that validates and builds python command line argument list
+     */
+    private List<String> validateAndBuildCommand(final String scriptName, final List<String> pythonProcessArgs, final List<String> scriptArgs) {
```
This method can be static.

cmnbroad reviewed (approved):
@asmirnov239 This looks pretty good now - only one minor comment. I'm not sure if this has already had a merge test run with the changes in https://github.com/broadinstitute/gatk/pull/6554, so it might be worth rebasing and running the tests one more time before merge.
